### ISSUE TITLE
[FEATURE]: Storing values larger than 2 GB

### ISSUE THREAD

**Issue by mitar** [dgraph-io/badger#2040]
### Have you tried Badger before this proposal? and did not find anything similar?

None

### What you wanted to do.

I wanted to set `ValueLogFileSize` to 6 GB.

### What you actually did.

Called `opts.WithValueLogFileSize(6 * 1024 * 1024 * 1024)`.

### Why wasn't it great, with examples.

I got error `Invalid ValueLogFileSize, must be in range [1MB, 2GB)`.

### Additional information.

Could this limit be lifted on 64bit systems? It seems to me that this is just to support mmap on 32 bit systems and there is really no design limitation why not to support larger values? On my system, I have:

```
$ grep -m1 "address sizes" /proc/cpuinfo 
address sizes	: 39 bits physical, 48 bits virtual
```

So a bit less than 2^47 (128 TiB) could be the max limit for mmap on my system.

**Comment by simon28082** [comment]
I have the same problem, 2G is too little for me, we need more than 8G

**Comment by nathanhack** [comment]
Same here

**Comment by mitar** [comment]
Still relevant.

**Comment by mitar** [comment]
Still relevant.

**Comment by sufan721** [comment]
Indeed, the limit should be raised, or users should be allowed to choose the limit.

**Comment by mitar** [comment]
Still relevant.

**Comment by matthewmcneely** [comment]
Some context on why this limit isn't a simple validation tweak: the 2GB cap is baked into Badger's on-disk format.

The value pointers stored in the LSM tree (`valuePointer` in [structs.go](https://github.com/dgraph-io/badger/blob/main/structs.go)) hold `Fid`, `Len`, and `Offset` as `uint32`, and the vlog entry header's value length is also decoded into a `uint32`. That caps addressable offsets in a single vlog file at 4GB. The reason the setting stops at 2GB rather than 4GB is that rotation is lazy: Badger only rolls to a new vlog file *after* the write offset exceeds `ValueLogFileSize`, and a single value can itself be up to `ValueLogFileSize`, so files are created and mmapped at `2*ValueLogFileSize` ([value.go](https://github.com/dgraph-io/badger/blob/main/value.go)). The worst-case file size approaches 2x the setting, which must stay under `uint32` max, hence the `< 2^31` bound. Supporting 32-bit builds was part of the original reasoning as well.

This came up on the Dgraph forums a couple of times:

- [Is it possible to change value pointer offset uint32 to uint64?](https://discuss.dgraph.io/t/is-it-possible-to-change-value-pointer-offset-uint32-to-uint64/10943) — a Badger maintainer confirmed the mmap/32-bit rationale and that widening to `uint64` is possible on 64-bit machines, "but you would have to make these changes a bunch of places."
- [Storing blobs in Dgraph](https://discuss.dgraph.io/t/storing-blobs-in-dgraph/1927) — discussion of the ~4GB theoretical ceiling from the pointer format and the deliberate 2GB cap, with the recommendation to keep large blobs outside the store.

So lifting the limit means widening the value pointer and vlog header fields, which changes both the vlog layout and the pointers persisted inside every SST. That's a disk-format version bump plus a migration path, not just relaxing the check in `Open`. It would also have limited practical payoff for very large values as things stand today: values aren't streamed anywhere in the write path, read path, or vlog GC, so a 6GB value would mean 6GB+ allocations on every touch.

For values larger than the cap, chunking across multiple keys (or storing blobs externally and keeping a reference in Badger) should work well and is what the maintainers have historically recommended.
